cpp/src/lists/contains.cu

```diff
@@ -209,31 +329,87 @@ struct lookup_functor {
                                  mr);
     }
 
-    auto const device_view = column_device_view::create(lists.parent(), stream);
-    auto const d_lists     = lists_column_device_view{*device_view};
-    auto const d_skeys     = get_search_keys_device_iterable_view(search_key, stream);
+    auto const lists_cdv_ptr = column_device_view::create(lists.parent(), stream);
+    auto const keys_dv_ptr   = get_search_keys_device_view_ptr(search_keys, stream);
 
-    auto result_positions = make_numeric_column(
+    auto out_positions = make_numeric_column(
       data_type{type_id::INT32}, lists.size(), cudf::mask_state::UNALLOCATED, stream, mr);
-    auto result_validity = make_numeric_column(
-      data_type{type_id::BOOL8}, lists.size(), cudf::mask_state::UNALLOCATED, stream, mr);
-    auto mutable_result_positions =
-      mutable_column_device_view::create(result_positions->mutable_view(), stream);
-    auto mutable_result_validity =
-      mutable_column_device_view::create(result_validity->mutable_view(), stream);
-    auto search_key_iter =
-      cudf::detail::make_pair_rep_iterator<ElementType, search_keys_have_nulls>(*d_skeys);
-
-    search_each_list_row<ElementType>(d_lists,
-                                      search_key_iter,
-                                      find_option,
-                                      *mutable_result_positions,
-                                      *mutable_result_validity,
-                                      stream);
-
-    auto [null_mask, num_nulls] = construct_null_mask(lists, result_validity->view(), stream, mr);
-    result_positions->set_null_mask(std::move(null_mask), num_nulls);
-    return result_positions;
+    auto out_validity   = rmm::device_uvector<bool>(lists.size(), stream);
+    auto const out_iter = thrust::make_zip_iterator(
+      out_positions->mutable_view().template begin<size_type>(), out_validity.begin());
+    auto const searcher = search_functor<Type>{};
+
+    if constexpr (std::is_same_v<Type, cudf::struct_view>) {
+#if 1
+      // Prepare to flatten the structs column and scalar.
+      auto const child_tview   = table_view{{child}};
+      auto const keys_tview    = get_search_keys_table_view(search_keys);
+      auto const has_any_nulls = has_nested_nulls(child_tview) || has_nested_nulls(keys_tview);
+      auto const flatten_nullability = has_any_nulls
+                                         ? structs::detail::column_nullability::FORCE
+                                         : structs::detail::column_nullability::MATCH_INCOMING;
+
+      // Flatten the input search space column and the keys scalar/column, only materialize the
+      // bitmask if there is null in the input.
+      auto const child_flattened =
+        structs::detail::flatten_nested_columns(child_tview, {}, {}, flatten_nullability);
+      auto const keys_flattened =
+        structs::detail::flatten_nested_columns(keys_tview, {}, {}, flatten_nullability);
```

Comment:
This flattening logic should not be necessary anymore with the new row_comparators.

Reply:
But it is still in `experimental::` namespace, isn't it?

Reply:
Struct flattening has already been used in many places (sorting/min/max/contains etc...). I planned to use this flattening old way for all the ongoing FEA (`lists::contains` and `cudf::contains`), then make a separate, big PR to convert all these to use the new experimental row comparator.

Reply:
The namespace is used for the purpose of coexisting the old comparators and the new comparators to facilitate a gradual transition. 

If the new row comparators will make this PR simpler while adding functionality, you can just use it without worrying much about stability.

Reply:
Okay got it. Thanks.

Reply:
Also, no need to make a large big bang PR that changes over all contains logic to use the new comparators. You can take it one step at a time. 

Since the original issue #8958 does request deeply nested structs, you would get that functionality for "free".

Reply:
So, based on our discussions, we still don't have eq comparator for 2 structs columns yet. I'll just use the old flattening way for this PR. A complete conversion to use the new comparator will be done later.

Reply:
> I'll just use the old flattening way for this PR. A complete conversion to use the new comparator will be done later.

Let's not add throwaway work of building new functionality on top of something we know we want to get rid of.

Reply:
Okay. So @bdice I'm relying on you :smirk: